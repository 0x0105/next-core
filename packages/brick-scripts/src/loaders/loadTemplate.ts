import fs from "fs-extra";
import os from "os";
import path from "path";
import klawSync from "klaw-sync";
import * as changeCase from "change-case";
import rp from "request-promise-native";
import { FileWithContent, TargetType } from "../interface";
import { getEasyopsConfig } from "../getEasyopsConfig";
const { exec } = require("child_process");
// `tsc` will compile files which `import` or `require`,
// thus, we read file content instead of importing.
const easyopsConfig = getEasyopsConfig();
const packageJson = JSON.parse(
  fs.readFileSync(path.join(__dirname, "../../package.json"), "utf8")
);
const { templateDependencies } = packageJson;
const basicBricksVersion = templateDependencies["@bricks/basic-bricks"];
const brickContainerVersion = templateDependencies["@easyops/brick-container"];

const workspacePackageJson = JSON.parse(
  fs.readFileSync(path.join(process.cwd(), "package.json"), "utf8")
);
const workspaceHomepage = workspacePackageJson.homepage;
let userName = "bot";
exec("git config user.name", { encoding: "utf8" }, function (err, value) {
  if (err) return;
  userName = value.trim();
});

function escapeRegExp(str: string): string {
  return str.replace(/([.*+?^=!:${}()|[\]/\\])/g, "\\$1");
}

function replaceAll(str: string, find: string, replace: string): string {
  return str.replace(new RegExp(escapeRegExp(find), "g"), replace);
}

function replaceFileContent(
  filePath: string,
  translations: Record<string, string>
): string {
  const content = fs.readFileSync(filePath, "utf8");
  const contentFinds = Object.keys(translations).filter((key) =>
    content.includes(key)
  );
  if (contentFinds.length > 0) {
    return contentFinds.reduce(
      (acc, find) => replaceAll(acc, find, translations[find]),
      content
    );
  }
  return content;
}

function replaceDepsVersion(jsonString: string): string {
  const pkg = JSON.parse(jsonString);

  const peerDeps = pkg.peerDependencies || {};
  if (peerDeps["@bricks/basic-bricks"]) {
    peerDeps["@bricks/basic-bricks"] = basicBricksVersion;
  }
  return JSON.stringify(pkg, null, 2) + os.EOL;
}

export async function loadTemplate({
  targetType,
  packageName,
  brickName,
  templateName,
  processorName,
  targetRoot,
  docRoot,
}: {
  targetType: TargetType;
  packageName: string;
  brickName: string;
  templateName: string;
  processorName: string;
  targetRoot: string;
  docRoot: string;
}): Promise<FileWithContent[]> {
  const targetMap: { [key: string]: string } = {
    [TargetType.A_NEW_BRICK]: "brick",
    [TargetType.A_NEW_CUSTOM_TEMPLATE]: "custom-template",
    [TargetType.A_NEW_PACKAGE_OF_BRICKS]: "bricks-pkg",
    [TargetType.A_NEW_PACKAGE_OF_LIBS]: "libs-pkg",
    [TargetType.A_NEW_PACKAGE_OF_MICRO_APPS]: "micro-apps-pkg",
    [TargetType.A_NEW_CUSTOM_PROVIDER_BRICK]: "custom-provider-brick",
    [TargetType.A_NEW_CUSTOM_PROCESSOR]: "custom-processor",
    [TargetType.A_NEW_PACKAGE_OF_PROVIDERS]: "providers-pkg",
    [TargetType.A_NEW_PACKAGE_OF_DLL]: "dll-pkg",
    [TargetType.TRANSFORM_A_MICRO_APP]: "transformed-micro-apps-pkg",
    [TargetType.A_NEW_LEGACY_TEMPLATE]: "template",
    [TargetType.A_NEW_PACKAGE_OF_LEGACY_TEMPLATES]: "templates-pkg",
    [TargetType.I18N_PATCH_A_PACKAGE_OF_LEGACY_TEMPLATES]:
      "i18n-patched-templates-pkg",
  };
  const templateDirOfFileName = targetMap[targetType];
  const templateRoot = path.join(__dirname, "../../template");
  let templateDir = path.join(templateRoot, templateDirOfFileName);

  const ignores = [".DS_Store"];
  let sdkName: string;
  if (targetType === TargetType.A_NEW_PACKAGE_OF_PROVIDERS) {
    // Providers 库直接使用构件库的模板文件，但忽略其 `src` 目录。
    templateDir = path.join(
      templateRoot,
      targetMap[TargetType.A_NEW_PACKAGE_OF_BRICKS]
    );
    ignores.push("template/bricks-pkg/src/");
    sdkName = `${packageName.replace(/^providers-of-/, "")}-sdk`;
  }

  const translations: Record<string, string> = {
    "$workspace-homepage$": workspaceHomepage,
    "$kebab-package-name$": packageName,
    $PascalPackageName$: changeCase.pascalCase(packageName),
    "$Title Package Name$": changeCase.capitalCase(packageName),
    $CONSTANT_PACKAGE_NAME$: changeCase.constantCase(packageName),
    $camelCasePackageName$: changeCase.camelCase(packageName),
    $PascalBrickName$: changeCase.pascalCase(brickName),
    "$kebab-brick-name$": `${packageName}.${brickName}`,
    "$kebab-brick-last-name$": brickName,
    "$kebab-custom-provider-brick-name$": `${packageName}.provider-${brickName}`,
    "$generator.version$": `v${packageJson.version}`,
    "$brick.container.version$": brickContainerVersion,
    "$kebab-sdk-name$": sdkName,
    "$kebab-template-name$": templateName,
    "$kebab-username$": userName,
    $camelTemplateName$: changeCase.camelCase(templateName),
    $PascalTemplateName$: changeCase.pascalCase(templateName),
    $camelProcessorName$: changeCase.camelCase(processorName),
  };

  const filter = (src: string): boolean =>
    ignores.every((item) => !src.includes(item));

  const templateGroups: any = [
    {
      templateDir,
      targetDir: targetRoot,
      files: klawSync(templateDir, {
        depthLimit: 4,
        nodir: true,
        filter: (item) => filter(item.path),
      }),
    },
  ];

  if (targetType === TargetType.A_NEW_PACKAGE_OF_BRICKS && brickName) {
    // Also create a new brick for the new bricks-package
    const brickTemplateDir = path.join(templateRoot, "brick");
    templateGroups.push({
      templateDir: brickTemplateDir,
      targetDir: path.join(targetRoot, "src", brickName),
      files: klawSync(brickTemplateDir, {
        depthLimit: 2,
        nodir: true,
        filter: (item) => filter(item.path),
      }),
    });
  } else if (targetType === TargetType.A_NEW_PACKAGE_OF_LEGACY_TEMPLATES) {
    // Also create a new brick for the new bricks-package
    const templateTemplateDir = path.join(templateRoot, "template");
    templateGroups.push({
      templateDir: templateTemplateDir,
      targetDir: path.join(targetRoot, "src"),
      files: klawSync(templateTemplateDir, {
        depthLimit: 1,
        nodir: true,
        filter: (item) => filter(item.path),
      }),
    });
  } else if (targetType === TargetType.A_NEW_PACKAGE_OF_PROVIDERS) {
    // Providers 库还有专属的一些模板文件。
    const brickTemplateDir = path.join(templateRoot, "providers-pkg");
    templateGroups.push({
      templateDir: brickTemplateDir,
      targetDir: targetRoot,
      files: klawSync(brickTemplateDir, {
        depthLimit: 2,
        nodir: true,
        filter: (item) => filter(item.path),
      }),
    });

    let sdkVersion;
    const getVersion = async () => {
      try {
        const sdkPackage = await rp({
          url: `https://registry.npm.easyops.local/@sdk/${sdkName}`,
          json: true,
          strictSSL: false,
        });
        sdkVersion = `^${sdkPackage["dist-tags"].latest}`;
      } catch {
        sdkVersion = "FETCH LATEST VERSION ERROR";
      }
      translations["$sdk.version$"] = sdkVersion;
    };

    if (!easyopsConfig?.useLocalSdk) {
      await getVersion();
    } else {
      translations["$sdk.version$"] = "1.0.0";
    }
  }

  const files: FileWithContent[] = templateGroups.reduce(
    (acc: FileWithContent[], group: any) =>
      acc.concat(
        group.files.map((file: any) => {
          const content = replaceFileContent(file.path, translations);

          let realPath = file.path;
          const basename = path.basename(file.path);
          const filenameFinds = Object.keys(translations).filter((key) =>
            basename.includes(key)
          );
          if (filenameFinds.length > 0) {
            realPath = path.join(
              path.dirname(realPath),
              filenameFinds.reduce(
                (acc, find) => replaceAll(acc, find, translations[find]),
                basename
              )
            );
          }

          return [
            path.join(
              group.targetDir,
              path.relative(group.templateDir, realPath)
            ),
            content,
          ];
        })
      ),
    []
  );

  if (
    targetType !== TargetType.A_NEW_BRICK &&
    targetType !== TargetType.A_NEW_CUSTOM_TEMPLATE &&
    targetType !== TargetType.A_NEW_CUSTOM_PROVIDER_BRICK &&
    targetType !== TargetType.A_NEW_CUSTOM_PROCESSOR &&
    targetType !== TargetType.A_NEW_LEGACY_TEMPLATE &&
    targetType !== TargetType.TRANSFORM_A_MICRO_APP &&
    targetType !== TargetType.I18N_PATCH_A_PACKAGE_OF_LEGACY_TEMPLATES
  ) {
    files.push([
      path.join(targetRoot, "package.json"),
      replaceDepsVersion(
        replaceFileContent(
          path.join(templateRoot, `${targetMap[targetType]}.json`),
          translations
        )
      ),
    ]);
  }

  return files;
}
