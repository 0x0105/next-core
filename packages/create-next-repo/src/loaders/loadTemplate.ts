import fs from "fs";
import os from "os";
import path from "path";
import klawSync from "klaw-sync";
import * as changeCase from "change-case";

type FileWithContent = [string, string];

// `tsc` will compile files which `import` or `require`,
// thus, we read file content instead of importing.
const packageJson = JSON.parse(
  fs.readFileSync(path.join(__dirname, "../../package.json"), "utf8")
);
const devDependenciesCopyMap = {
  devDependencies: [
    "@easyops/babel-preset-next",
    "@easyops/brick-container",
    "@easyops/brick-scripts",
    "@easyops/browserslist-config-next",
    "@easyops/build-config-factory",
    "@easyops/eslint-config-next",
    "@easyops/webpack-config-factory"
  ],
  templateDependencies: ["@bricks/basic-bricks"]
};

function escapeRegExp(str: string): string {
  return str.replace(/([.*+?^=!:${}()|[\]/\\])/g, "\\$1");
}

function replaceAll(str: string, find: string, replace: string): string {
  return str.replace(new RegExp(escapeRegExp(find), "g"), replace);
}

function replaceFileContent(
  filePath: string,
  translations: Record<string, string>
): string {
  const content = fs.readFileSync(filePath, "utf8");
  const contentFinds = Object.keys(translations).filter(key =>
    content.includes(key)
  );
  if (contentFinds.length > 0) {
    return contentFinds.reduce(
      (acc, find) => replaceAll(acc, find, translations[find]),
      content
    );
  }
  return content;
}

function replaceDepsVersion(jsonString: string): string {
  const pkg = JSON.parse(jsonString);
  for (const [type, deps] of Object.entries(devDependenciesCopyMap)) {
    for (const dep of deps) {
      pkg.devDependencies[dep] = packageJson[type][dep];
    }
  }
  return JSON.stringify(pkg, null, 2) + os.EOL;
}

export function loadTemplate(
  repoName: string,
  targetDir: string,
  flags: { internal: boolean }
): FileWithContent[] {
  const templateDir = path.join(__dirname, "../../template");

  const ignores = [".DS_Store"];
  const filter = (src: string): boolean =>
    ignores.some(item => !src.includes(item));

  const templateFiles = klawSync(templateDir, {
    depthLimit: 3,
    nodir: true,
    filter: item => filter(item.path)
  });

  const translations: Record<string, string> = {
    "$kebab-repo-name$": changeCase.paramCase(repoName),
    "$Title Repo Name$": changeCase.capitalCase(repoName),
    "$generator.version$": `v${packageJson.version}`,
    "$easyops-registry$": flags.internal
      ? "http://registry.npm.easyops.local"
      : "http://r.pnpm.easyops.cn"
  };

  const files: FileWithContent[] = templateFiles.map(file => {
    const content = replaceFileContent(file.path, translations);
    return [
      path.join(targetDir, path.relative(templateDir, file.path)),
      content
    ];
  });

  files.push([
    path.join(targetDir, "package.json"),
    replaceDepsVersion(
      replaceFileContent(
        path.join(templateDir, "../template.json"),
        translations
      )
    )
  ]);

  return files;
}
