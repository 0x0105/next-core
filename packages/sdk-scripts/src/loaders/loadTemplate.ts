import fs from "fs";
import os from "os";
import path from "path";
import klawSync from "klaw-sync";
import changeCase from "change-case";
import { FileWithContent } from "../interface";

// `tsc` will compile files which `import` or `require`,
// thus, we read file content instead of importing.
const packageJson = JSON.parse(
  fs.readFileSync(path.join(__dirname, "../../package.json"), "utf8")
);
const { devDependencies } = packageJson;
const brickHttpVersion = devDependencies["@easyops/brick-http"];
const rollupConfigFactoryVersion =
  devDependencies["@easyops/rollup-config-factory"];

function escapeRegExp(str: string): string {
  return str.replace(/([.*+?^=!:${}()|[\]/\\])/g, "\\$1");
}

function replaceAll(str: string, find: string, replace: string): string {
  return str.replace(new RegExp(escapeRegExp(find), "g"), replace);
}

function replaceFileContent(
  filePath: string,
  translations: Record<string, string>
): string {
  const content = fs.readFileSync(filePath, "utf8");
  const contentFinds = Object.keys(translations).filter(key =>
    content.includes(key)
  );
  if (contentFinds.length > 0) {
    return contentFinds.reduce(
      (acc, find) => replaceAll(acc, find, translations[find]),
      content
    );
  }
  return content;
}

function replaceDepsVersion(jsonString: string, sdkVersion: string): string {
  const pkg = JSON.parse(jsonString);
  pkg.devDependencies["@easyops/brick-http"] = brickHttpVersion;
  pkg.devDependencies[
    "@easyops/rollup-config-factory"
  ] = rollupConfigFactoryVersion;
  if (sdkVersion) {
    pkg.version = sdkVersion;
  }
  return JSON.stringify(pkg, null, 2) + os.EOL;
}

export function loadTemplate(
  serviceName: string,
  sdkRoot: string,
  sdkVersion: string
): FileWithContent[] {
  const templateDir = path.join(__dirname, "../../template");

  const ignores = [".DS_Store"];
  const filter = (src: string): boolean =>
    ignores.some(item => !src.includes(item));

  const templateFiles = klawSync(templateDir, {
    depthLimit: 2,
    nodir: true,
    filter: item => filter(item.path)
  });

  const translations: Record<string, string> = {
    "$kebab-service-name$": changeCase.kebab(serviceName),
    "$Title Service Name$": changeCase.title(serviceName),
    $PascalServiceName$: changeCase.pascal(serviceName),
    "$generator.version$": `v${packageJson.version}`
  };

  const files: FileWithContent[] = templateFiles.map(file => {
    const content = replaceFileContent(file.path, translations);
    return [path.join(sdkRoot, path.relative(templateDir, file.path)), content];
  });

  files.push([
    path.join(sdkRoot, "package.json"),
    replaceDepsVersion(
      replaceFileContent(
        path.join(templateDir, "../template.json"),
        translations
      ),
      sdkVersion
    )
  ]);

  return files;
}
