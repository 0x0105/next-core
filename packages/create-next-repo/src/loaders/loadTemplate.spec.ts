import path from "path";
import { loadTemplate } from "./loadTemplate";

describe("loadTemplate", () => {
  it("should work", () => {
    let files = loadTemplate("my-repo", "/tmp/next/my-repo");
    files = files.filter(([filePath]) => {
      const fileName = path.basename(filePath);
      return !["package.json", "README.md"].includes(fileName);
    });
    expect(files).toMatchSnapshot();
  });
});
