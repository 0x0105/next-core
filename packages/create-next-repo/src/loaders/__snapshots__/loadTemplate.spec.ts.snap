// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`loadTemplate should work by default 1`] = `
Array [
  Array [
    "/tmp/next/my-repo/.editorconfig",
    "root = true

[*]
indent_style = space
indent_size = 2
charset = utf-8
trim_trailing_whitespace = true
insert_final_newline = true

[*.{md,snap}]
trim_trailing_whitespace = false
",
  ],
  Array [
    "/tmp/next/my-repo/.eslintignore",
    "node_modules
dist
",
  ],
  Array [
    "/tmp/next/my-repo/.eslintrc",
    "{
  \\"root\\": true,
  \\"extends\\": \\"@easyops/eslint-config-next\\"
}
",
  ],
  Array [
    "/tmp/next/my-repo/.gitignore",
    "dist
node_modules
.coverage
.vscode
.idea
.DS_Store
.schema
*.log
report.*.json
/dev-settings.yaml
/dev.config.js

/bricks/*/deploy/*
!/bricks/*/deploy/package.conf.yaml
/bricks/*/.pkgbuild/
/bricks/*/version.ini
/bricks/*/*.tgz

/micro-apps/*/deploy/
/micro-apps/*/.pkgbuild/
/micro-apps/*/version.ini
/micro-apps/*/*.tgz

/templates/*/deploy/
/templates/*/.pkgbuild/
/templates/*/version.ini
/templates/*/*.tgz

",
  ],
  Array [
    "/tmp/next/my-repo/.huskyrc",
    "{
  \\"hooks\\": {
    \\"pre-commit\\": \\"lint-staged\\"
  }
}
",
  ],
  Array [
    "/tmp/next/my-repo/.lintstagedrc",
    "{
  \\"*.{js,ts,jsx,tsx}\\": [
    \\"cross-env ESLINT_ENV='production' eslint --fix --quiet\\",
    \\"prettier --write\\",
    \\"git add\\"
  ],
  \\"*.css\\": [
    \\"prettier --write\\",
    \\"git add\\"
  ],
  \\"*.html\\": [
    \\"prettier --write\\",
    \\"git add\\"
  ],
  \\"*.md\\": [
    \\"prettier --write\\",
    \\"git add\\"
  ],
  \\"*.json\\": [
    \\"prettier --write\\",
    \\"git add\\"
  ]
}
",
  ],
  Array [
    "/tmp/next/my-repo/.npmrc",
    "registry=https://registry.npm.taobao.org
@easyops:registry=http://r.pnpm.easyops.cn
@bricks:registry=http://r.pnpm.easyops.cn
@dll:registry=http://r.pnpm.easyops.cn
@libs:registry=http://r.pnpm.easyops.cn
@micro-apps:registry=http://r.pnpm.easyops.cn
@sdk:registry=http://r.pnpm.easyops.cn
@templates:registry=http://r.pnpm.easyops.cn
tag-version-prefix=
",
  ],
  Array [
    "/tmp/next/my-repo/.pkgbuild/@bricks/PKGBUILD",
    "# The common PKGBUILD of all @bricks/*
# Keep in mind that the actual PKGBUILD execution context is @bricks/[plugin-name]
pkgver() {
  jq -r '.version' < \\"\${srcdir}/src/package.json\\"
}
arch=('x86_64')
pkgrel=1
source=()
options=()

_prepare_source_links() {
  local _projectdir=\\"\${startdir}/..\\"
  ln -snf \\"\${_projectdir}\\" \\"\${srcdir}/src\\"
}

_package_artifacts() {
  rsync --archive --stats --include=\\"dist/***\\" --include=\\"deploy/***\\" --include=\\"CHANGELOG.md\\" --include=\\"README.md\\" --exclude=\\"*\\" . \\"\${pkgdir}\\"
}

prepare() {
  _prepare_source_links
}

package() {
  cd \\"\${srcdir}/src\\"
  _package_artifacts
}
",
  ],
  Array [
    "/tmp/next/my-repo/.pkgbuild/@micro-apps/PKGBUILD",
    "# The common PKGBUILD of all @bricks/*
# Keep in mind that the actual PKGBUILD execution context is @bricks/[plugin-name]
arch=('x86_64')
pkgrel=1
source=()
options=()

_prepare_source_links() {
  local _projectdir=\\"\${startdir}/..\\"
  ln -snf \\"\${_projectdir}\\" \\"\${srcdir}/src\\"
}

_package_artifacts() {
  rsync --archive --stats --exclude=\\"package.json\\" --exclude=\\"scripts\\" . \\"\${pkgdir}\\"
}

prepare() {
  _prepare_source_links
}

pkgver() {
  jq -r '.version' < \\"\${srcdir}/src/package.json\\"
}

package() {
  cd \\"\${srcdir}/src\\"
  _package_artifacts
}
",
  ],
  Array [
    "/tmp/next/my-repo/.pkgbuild/@templates/PKGBUILD",
    "# The common PKGBUILD of all @templates/*
# Keep in mind that the actual PKGBUILD execution context is @templates/[plugin-name]
pkgver() {
  jq -r '.version' < \\"\${srcdir}/src/package.json\\"
}
arch=('x86_64')
pkgrel=1
source=()
options=()

_prepare_source_links() {
  local _projectdir=\\"\${startdir}/..\\"
  ln -snf \\"\${_projectdir}\\" \\"\${srcdir}/src\\"
}

_package_artifacts() {
  rsync --archive --stats --include=\\"dist/***\\" --include=\\"deploy/***\\" --include=\\"CHANGELOG.md\\" --include=\\"README.md\\" --exclude=\\"*\\" . \\"\${pkgdir}\\"
}

prepare() {
  _prepare_source_links
}

package() {
  cd \\"\${srcdir}/src\\"
  _package_artifacts
}
",
  ],
  Array [
    "/tmp/next/my-repo/.size-limit.js",
    "const { sizeLimit } = require(\\"@easyops/build-config-factory\\");

module.exports = sizeLimit({
  bricks: {
    \\"*\\": \\"50 KB\\"
  },
  templates: {
    \\"*\\": \\"5 KB\\"
  }
});
",
  ],
  Array [
    "/tmp/next/my-repo/.yarnrc",
    "registry \\"https://registry.npm.taobao.org\\"
\\"@easyops:registry\\" \\"http://r.pnpm.easyops.cn\\"
\\"@bricks:registry\\" \\"http://r.pnpm.easyops.cn\\"
\\"@dll:registry\\" \\"http://r.pnpm.easyops.cn\\"
\\"@libs:registry\\" \\"http://r.pnpm.easyops.cn\\"
\\"@micro-apps:registry\\" \\"http://r.pnpm.easyops.cn\\"
\\"@sdk:registry\\" \\"http://r.pnpm.easyops.cn\\"
\\"@templates:registry\\" \\"http://r.pnpm.easyops.cn\\"
",
  ],
  Array [
    "/tmp/next/my-repo/__jest__/customElementsV1Serializer.ts",
    "// 因 jest 引入 document-register-element 后，文档中的 Element 对象的 constructor 变成了 CustomElementsV1，这里将该类型的 serialize 重新指回 pretty-format 的 DOMElement.serialize，以便正常序列化
import { plugins } from \\"pretty-format\\";

export const test = (val: any) =>
  val &&
  val.constructor &&
  val.constructor.name &&
  val.constructor.name === \\"CustomElementsV1\\";
export const serialize = plugins.DOMElement.serialize;
",
  ],
  Array [
    "/tmp/next/my-repo/__jest__/setup.ts",
    "import { configure } from \\"enzyme\\";
import Adapter from \\"enzyme-adapter-react-16\\";
import { act } from \\"react-dom/test-utils\\";

configure({ adapter: new Adapter() });

// Ref https://github.com/facebook/jest/issues/2157#issuecomment-279171856
(global as any).flushPromises = () =>
  act(() => new Promise(resolve => setImmediate(resolve)));

Element.prototype.scrollIntoView = jest.fn();
document.execCommand = jest.fn(() => true);
",
  ],
  Array [
    "/tmp/next/my-repo/__mocks__/htmlMock.js",
    "export default \\"test-file-stub\\";
",
  ],
  Array [
    "/tmp/next/my-repo/__mocks__/markdownMock.js",
    "export default \\"markdown mock\\";
",
  ],
  Array [
    "/tmp/next/my-repo/__mocks__/pngMock.js",
    "export default \\"png-url\\";
",
  ],
  Array [
    "/tmp/next/my-repo/__mocks__/react-i18next.js",
    "export const withTranslation = () => Component => {
  Component.defaultProps = {
    ...Component.defaultProps,
    t: key => key,
    i18n: {
      language: \\"zh-CN\\"
    }
  };
  return Component;
};

export const useTranslation = () => ({
  t: key => key,
  i18n: {
    language: \\"zh-CN\\"
  }
});
",
  ],
  Array [
    "/tmp/next/my-repo/__mocks__/styleMock.js",
    "export default {};
",
  ],
  Array [
    "/tmp/next/my-repo/__mocks__/svgrMock.js",
    "export default \\"svgr-url\\";
export const ReactComponent = \\"svgr-mock\\";
",
  ],
  Array [
    "/tmp/next/my-repo/babel.config.js",
    "module.exports = {
  presets: [\\"@easyops/babel-preset-next\\"]
};
",
  ],
  Array [
    "/tmp/next/my-repo/declarations/global.d.ts",
    "declare module \\"*.module.css\\" {
  const classes: { [key: string]: string };
  export default classes;
}

declare module \\"*.css\\" {
  const css: string;
  export default css;
}

declare module \\"*.less\\" {
  const lessValue: string;
  export default lessValue;
}

declare module \\"*.html\\" {
  const html: string;
  export default html;
}

declare module \\"*.md\\" {
  const content: string;
  export default content;
}

interface SvgrComponent
  extends React.StatelessComponent<React.SVGAttributes<SVGElement>> {}

declare module \\"*.svg\\" {
  const url: string;
  export default url;
  export const ReactComponent: SvgrComponent;
}

declare module \\"*.png\\" {
  const url: string;
  export default url;
}

declare namespace JSX {
  interface IntrinsicElements {
    slot: any;
  }
}
",
  ],
  Array [
    "/tmp/next/my-repo/jest.config.js",
    "module.exports = {
  setupFilesAfterEnv: [\\"<rootDir>/__jest__/setup.ts\\"],
  snapshotSerializers: [
    \\"enzyme-to-json/serializer\\",
    \\"<rootDir>/__jest__/customElementsV1Serializer.ts\\"
  ],
  coveragePathIgnorePatterns: [
    \\"/node_modules/\\",
    \\"/dist/\\",
    \\"\\\\\\\\.spec\\\\\\\\.[jt]sx?$\\",
    \\"/scripts/\\",
    \\"/__jest__/\\",
    \\"/micro-apps/\\"
  ],
  collectCoverage: true,
  coverageThreshold: {
    global: {
      statements: 80,
      branches: 50,
      functions: 80,
      lines: 80
    }
  },
  coverageDirectory: \\"<rootDir>/.coverage\\",
  coverageReporters: [\\"lcov\\", \\"text-summary\\"],
  transform: {
    \\"^.+\\\\\\\\.[jt]sx?$\\": \\"babel-jest\\"
  },
  testPathIgnorePatterns: [\\"/node_modules/\\", \\"/dist/\\"],
  moduleNameMapper: {
    \\"\\\\\\\\.module\\\\\\\\.css$\\": \\"identity-obj-proxy\\",
    \\"\\\\\\\\.less$\\": \\"<rootDir>/__mocks__/styleMock.js\\",
    \\"\\\\\\\\.css$\\": \\"<rootDir>/__mocks__/styleMock.js\\",
    \\"^.+\\\\\\\\.html$\\": \\"<rootDir>/__mocks__/htmlMock.js\\",
    \\"\\\\\\\\.svg\\": \\"<rootDir>/__mocks__/svgrMock.js\\",
    \\"\\\\\\\\.png\\": \\"<rootDir>/__mocks__/pngMock.js\\",
    \\"^.+\\\\\\\\.md?$\\": \\"<rootDir>/__mocks__/markdownMock.js\\",
    // Ref http://react-dnd.github.io/react-dnd/docs/testing#setup
    \\"^dnd-core$\\": \\"dnd-core/dist/cjs\\",
    \\"^react-dnd$\\": \\"react-dnd/dist/cjs\\",
    \\"^react-dnd-html5-backend$\\": \\"react-dnd-html5-backend/dist/cjs\\"
  },
  // Ref https://github.com/facebook/jest/issues/2070#issuecomment-431706685
  // Todo(steve): remove next line when issue fixed.
  modulePathIgnorePatterns: [\\"<rootDir>/.*/__mocks__\\"],
  // Use jsdom@14 which supports MutationObserver
  testEnvironment: \\"jest-environment-jsdom-fourteen\\",
  timers: \\"fake\\"
};
",
  ],
  Array [
    "/tmp/next/my-repo/lerna.json",
    "{
  \\"packages\\": [\\"bricks/*\\", \\"libs/*\\", \\"micro-apps/*\\", \\"templates/*\\"],
  \\"version\\": \\"independent\\",
  \\"npmClient\\": \\"yarn\\",
  \\"stream\\": true,
  \\"useWorkspaces\\": true,
  \\"conventionalCommits\\": true,
  \\"command\\": {
    \\"publish\\": {
      \\"npmClient\\": \\"npm\\",
      \\"allowBranch\\": \\"master\\",
      \\"verifyAccess\\": false,
      \\"message\\": \\"chore(release): publish\\",
      \\"registry\\": \\"https://registry.npm.easyops.local\\"
    }
  }
}
",
  ],
  Array [
    "/tmp/next/my-repo/renovate.json",
    "{
  \\"$schema\\": \\"http://json.schemastore.org/renovate\\",
  \\"extends\\": [\\"config:base\\"],
  \\"rangeStrategy\\": \\"bump\\",
  \\"semanticCommits\\": true,
  \\"commitMessagePrefix\\": \\"chore(deps):\\",
  \\"assignees\\": [\\"weareoutman\\"],
  \\"ignorePresets\\": [\\":ignoreModulesAndTests\\"],
  \\"packageRules\\": [
    {
      \\"depTypeList\\": [\\"devDependencies\\"],
      \\"automerge\\": true
    },
    {
      \\"packagePatterns\\": [
        \\"^@bricks/\\",
        \\"^@libs/\\",
        \\"^@micro-apps/\\",
        \\"^@sdk/\\",
        \\"^@templates/\\"
      ],
      \\"enabled\\": false
    },
    {
      \\"packagePatterns\\": [\\"^@dll/\\", \\"^@easyops/\\"],
      \\"automerge\\": false,
      \\"groupName\\": \\"next-core packages\\"
    }
  ]
}
",
  ],
  Array [
    "/tmp/next/my-repo/tsconfig.json",
    "{
  \\"compilerOptions\\": {
    \\"declaration\\": false,
    \\"sourceMap\\": true,
    \\"importHelpers\\": true,
    \\"moduleResolution\\": \\"node\\",
    \\"resolveJsonModule\\": true,
    \\"esModuleInterop\\": true,
    \\"target\\": \\"ESNext\\",
    \\"module\\": \\"ESNext\\",
    \\"skipLibCheck\\": true,
    \\"strict\\": true,
    \\"strictNullChecks\\": false,
    \\"jsx\\": \\"preserve\\",
    \\"emitDecoratorMetadata\\": true,
    \\"experimentalDecorators\\": true
  },
  \\"exclude\\": [\\"node_modules\\", \\"**/*.spec.ts\\", \\"dist/\\", \\"**/__mocks__/\\"]
}
",
  ],
]
`;

exports[`loadTemplate should work for internal 1`] = `
Array [
  Array [
    "/tmp/next/my-repo/.npmrc",
    "registry=https://registry.npm.taobao.org
@easyops:registry=http://registry.npm.easyops.local
@bricks:registry=http://registry.npm.easyops.local
@dll:registry=http://registry.npm.easyops.local
@libs:registry=http://registry.npm.easyops.local
@micro-apps:registry=http://registry.npm.easyops.local
@sdk:registry=http://registry.npm.easyops.local
@templates:registry=http://registry.npm.easyops.local
tag-version-prefix=
",
  ],
  Array [
    "/tmp/next/my-repo/.yarnrc",
    "registry \\"https://registry.npm.taobao.org\\"
\\"@easyops:registry\\" \\"http://registry.npm.easyops.local\\"
\\"@bricks:registry\\" \\"http://registry.npm.easyops.local\\"
\\"@dll:registry\\" \\"http://registry.npm.easyops.local\\"
\\"@libs:registry\\" \\"http://registry.npm.easyops.local\\"
\\"@micro-apps:registry\\" \\"http://registry.npm.easyops.local\\"
\\"@sdk:registry\\" \\"http://registry.npm.easyops.local\\"
\\"@templates:registry\\" \\"http://registry.npm.easyops.local\\"
",
  ],
]
`;
